internal/gps/source_manager.go

```diff
@@ -349,37 +332,20 @@ func (e CouldNotCreateLockError) Error() string {
 // longer safe to call methods against it; all method calls will immediately
 // result in errors.
 func (sm *SourceMgr) Release() {
-	// Set sm.releasing before entering the Once func to guarantee that no
-	// _more_ method calls will stack up if/while waiting.
-	atomic.CompareAndSwapInt32(&sm.releasing, 0, 1)
-
-	// Whether 'releasing' is set or not, we don't want this function to return
-	// until after the doRelease process is done, as doing so could cause the
-	// process to terminate before a signal-driven doRelease() call has a chance
-	// to finish its cleanup.
-	sm.relonce.Do(sm.doRelease)
-}
-
-// doRelease actually releases physical resources (files on disk, etc.).
-//
-// This must be called only and exactly once. Calls to it should be wrapped in
-// the sm.relonce sync.Once instance.
-func (sm *SourceMgr) doRelease() {
-	// Send the signal to the supervisor to cancel all running calls
+	// Send the signal to the supervisor to cancel all running calls.
 	sm.cancelAll()
 	sm.suprvsr.wait()
 
-	// Close the source coordinator.
-	sm.srcCoord.close()
-
-	// Close the file handle for the lock file and remove it from disk
-	sm.lf.Unlock()
-	os.Remove(filepath.Join(sm.cachedir, "sm.lock"))
+	if atomic.CompareAndSwapInt32(&sm.releasing, 0, 1) {
```

Comment:
this gets the ordering wrong - we're now marking ourselves as being in the process of releasing, and therefore rejecting new incoming calls, only _after_ we've already waited for running commands to terminate. we need to start rejecting new calls immediately, signal for the termination of all existing calls, then wait for them all to terminate.

the old implementation had both a `sm.releasing` and `sm.released` `int32`s for these purposes. i replaced `sm.released` with the `sync.Once()` as it was a cleaner way of achieving the wait.

Reply:
Understood - I think this is fixed now, though.